sdk/core/azure-core/src/main/java/com/azure/core/implementation/FileWriteSubscriber.java

```diff
@@ -0,0 +1,108 @@
+
+    // volatile ensures that writes to these fields by one thread will be immediately visible to other threads.
+    // An I/O pool thread will write to isWriting and read isCompleted,
+    // while another thread may read isWriting and write to isCompleted.
+    private volatile boolean isWriting = false;
+    private volatile boolean isCompleted = false;
+
+    private final AsynchronousFileChannel fileChannel;
+    private final AtomicLong position;
+    private final MonoSink<Void> emitter;
+
+    private Subscription subscription;
+
+    public FileWriteSubscriber(AsynchronousFileChannel fileChannel, long position, MonoSink<Void> emitter) {
+        this.fileChannel = fileChannel;
+        this.position = new AtomicLong(position);
+        this.emitter = emitter;
+    }
+
+    @Override
+    public void onSubscribe(Subscription s) {
+        // Only set the Subscription if one has not been previously set.
+        // Any additional Subscriptions will be cancelled.
+        if (Operators.validate(this.subscription, s)) {
+            subscription = s;
+
+            s.request(1);
+        }
+    }
+
+    @Override
+    public void onNext(ByteBuffer bytes) {
+        try {
+            if (isWriting) {
+                onError(new IllegalStateException("Received onNext while processing another write operation."));
+            } else {
+                write(bytes);
+            }
+        } catch (Throwable throwable) {
+            // If writing has an error, and it isn't caught, there is a possibility for it to deadlock the reactive
+            // stream. Catch the exception and propagate it manually so that doesn't happen.
+            onError(throwable);
+        }
+    }
+
+    private void write(ByteBuffer bytes) {
+        isWriting = true;
+
+        fileChannel.write(bytes, position.get(), bytes, new CompletionHandler<Integer, ByteBuffer>() {
+            @Override
+            public void completed(Integer result, ByteBuffer attachment) {
+                position.addAndGet(result);
+
+                if (bytes.hasRemaining()) {
+                    // If the entire ByteBuffer hasn't been written send it to be written again until it completes.
+                    write(bytes);
+                } else {
+                    isWriting = false;
+                    if (isCompleted) {
+                        emitter.success();
+                    } else {
+                        subscription.request(1);
+                    }
+                }
+            }
+
+            @Override
+            public void failed(Throwable exc, ByteBuffer attachment) {
+                onError(exc);
+            }
+        });
+    }
+
+    @Override
+    public void onError(Throwable throwable) {
+        isWriting = false;
+        subscription.cancel();
+        emitter.error(throwable);
+    }
+
+    @Override
+    public void onComplete() {
+        isCompleted = true;
+        if (!isWriting) {
```

Comment:
I will update this and add a test for it.

Reply:
I think there is a code path where isWriting == true when upstream calls onComplete
 

Upstream calls OnNext (last chunk) -> initiate async-write (isWriting=true) -> onNext returned-> upstream call onComplete (isWriting is still true)-> async-write completes (isWriting=false) -> call downstream onComplete

Reply:
@anuchandy is correct, given that this is using an asynchronous file write there is a valid possibility for `onComplete` to trigger while writing is still occurring and for it to be a valid reactive streams scenario. 

A concrete example of this would be `Mono.just(buffer).flux()` being passed into `FluxUtil.writeFile` and the write operation taking a non-negligible amount of time. Upon subscribing to the `Mono.just` `Flux` it'll trigger `onNext` with the `buffer` and then once `onNext` in `FileWriteSubscriber` returns `Mono.just` will trigger `onComplete` terminating the stream.